### DISCUSSION TITLE
MP4 video playback sample needed

### DISCUSSION THREAD

**Discussion by mills32** [pspdev/pspdev#73] [General]
Hi. We can make a lot of things without video playback in a game or homebrew, but I'd love to have a minimal sample showing how to play an mp4 using the built-in decoding capabilities of the psp (pspmpeg mp4 AVC + AAC audio).

I really tried to create a sample (from cooleyes pmplayer), but I could never compile it, there are always missing functions, hundreds of compiller errors... It uses so many different libs and headers. Also a function called sceMpegGetAvcNalAu can't be found anywhere, I hope it can be raplaced by sceMpegGetAvcAu.

I managed to compile a program which initializes and loads the needed prx files using kubridge, and that's it, I have no idea of how to play an mp4.

I wanted this because it is incredibly easy to create a compatible mp4 file that will play on the PSP menu  (the binary pmplayer release can also play them, even 60fps files work fine).

So does anyone know how to do this?

Thanks!

**Comment by tpimh** [comment]
Unrelated, but I got theora+vorbis to play on the PSP twice: [using theoraplay](https://github.com/tpimh/jewels-SDL2-PSP-test/tree/theoraplay) and [using sdlogv](https://github.com/tpimh/jewels-SDL2-PSP-test/tree/sdlogv). The performance sucks, on real hardware it's single digit FPS numbers even with extremely low quality (PPSSPP is a bit better).

So yes, hardware accelerated video would be very nice.

**Comment by mills32** [comment]
[MP4_Play.zip](https://github.com/user-attachments/files/15668124/MP4_Play.zip)

This "monstruosity" tries to use the functions from the last pmplayer advance, to play "Files/test.mp4". It does not work. 

On PPSSPP sceAudioSetFrequency fails. 
On real PSP another sce fails (sceMpegCreate or something), it doesn't matter, I think there is some kind of huge bug in that code. Maybe someone can make it work, and we can remove 90% of the code to make a  sample for the sdk.
Tested video is an mp4 AVC 480x272 60fps, it worked perfectly with the psp menu video player, and PMP2.0 (an older version of pmlayer advance).

I tried to use PMP2.0 source to create the sample, but it uses a modified ffmpeg,  I just got lost trying to isolate the mp4 load function.

We have pmp lib which only plays that PMP format (480x272 23 fps). Converters don't even work on my pc, and the format is kind of difficult to create and limited.

Hope someone can use this.

**Reply by mirh** [reply]
PMP supposedly [doesn't work](https://web.archive.org/web/20080103043206/http://jonny.leffe.dnsalias.com/) with newer firmware? And [AFAICT](https://github.com/hrydgard/ppsspp/blob/v1.20.3/Core/HLE/sceAudio.cpp#L351) there's something missing in PPSSPP (or maybe [this](https://github.com/dbeef/spelunky-psp/pull/63)?)

As for `sceMpegCreate` yes it's voodoo and very hard to work with, but it was recently [achieved](https://andrewaltimit.github.io/oasis-os/journal/07-psp-media-engine.html).

**Reply by tpimh** [reply]
So decoding 480x272 H.264 from local network over WPA2 Wi-Fi without cacheing to Memory Stick should be possible, right?

**Reply by mirh** [reply]
Are you trying to port moonlight to the psp? :)
Anyway, that seems [very much](https://github.com/AndrewAltimit/oasis-os/blob/main/docs/psp-me-decode-next-steps.md#current-status-h264-decode-working-2026-03-25) what they claim to be achieving.
`Mpeg_mode=4` should be pretty straightforward to use with low-res content, while mode 5 can handle even 480p but it's fussy.

**Reply by tpimh** [reply]
This is so cool compared to my software-only tests

**Reply by mirh** [reply]
More samples that included the unknown parameters [here](https://lukasz.dk/mirror/forums.ps2dev.org/viewtopic2e95-2.html?t=9927) and [here](https://lukasz.dk/mirror/forums.ps2dev.org/viewtopic0ffe.html?t=11650).

**Comment by sharkwouter** [comment]
The build in decoder is your best bet, but we don't really have documentation for it. You can find some media players that support it when you search on GitHub. If you find out how to use it, please let us know. We've been wanting to have a good example for this, but we don't really have people that know how to use it it seems.

**Reply by tpimh** [reply]
The two links above I think are pretty good examples (I have not compiled/tested them, but looks nice), the only problem is that they are under GPL. So unless the original author (cooleyes) agrees to contribute these examples under BSD terms, they can't be included with PSPSDK.
